### PR TITLE
Edit Post

### LINKED ISSUE #40: Edit a Post
As a author I would like to be able to modify my Posts so that I can correct mistakes or add additional content.

**Given** the user is viewing the Post list  
**When** they select the option to edit an Post  
**Then** the user should be directed to a form and given the ability to change the Post's information  

**Given** the user is finished updating the Post information  
**When** they click the `Save` button  
**Then** the updated Post should be saved to the database  
**And** the user should be redirected to the new Post's details page

**Given** the user has decided not to edit the Post  
**When** they click the `Cancel` button  
**Then** the user should be redirected back to the list page

Editable Information

* Title
* Content
* Category
* Header Image URL (optional)
* Publication Date (optional)

### PR DESCRIPTION
**Pull request by cjgochanour**
#### Changes Made
1. Added file `PostEdit` to `tabloid/client/src/components` directory.
2. Modified file `postManager` to include `putPost()` that takes a post object and sends a PUT request to the database.
3. Modified file `PostController` to include `Put(int id, Post post)` that will determine whether or not to query the database with the new post object
4. Modified file `PostRepository` and its controller to include `UpdatePost(Post post)` that will execute a SQL query to update the given post
5. Modified file `ApplicationViews` to include routing for `PostEdit` component
​
#### Steps to Review
1. Checkout this branch locally.
    ```
    git fetch --all
    git checkout cg-editPost
    ```
2. Open a new Terminal tab (⌘T) and navigate to the server directory.
3. Test app functionality.
    > Log in to Tabloid if not logged in already
    > View post list. There are now edit buttons on each post
    > Click an edit button to be navigated to /posts/edit/:id
    > The values of each input field should be filled with the post's details already
    > Modify these values to your specifications
    > Click save to see changes reflected in the post's detail view, or cancel to be redirected back to post list
4. View code file.
    > Confirm file modifications are present as indicated above.
    > Confirm no unused code or extraneous comments exist.

Closes #40

### CODE REVIEW

File: Tabloid/client/src/modules/postManager.js [hunk 1]
```diff
@@ -54,3 +54,16 @@ export const addPost = (post) => {
         });
     });
 };
+
+export const putPost = (post) => {
```
Comment (SaraHackle) [inline Tabloid/client/src/modules/postManager.js]: Just for consistency, maybe consider calling this editPost? or I can change mine. It's up to you guys!

File: Tabloid/client/src/modules/postManager.js [hunk 2]
```diff
@@ -54,3 +54,16 @@ export const addPost = (post) => {
...
+    return getToken().then((token) => {
+        return fetch(`${_apiUrl}/${post.id}`, {
+            method: "PUT",
+            headers: {
+                Authorization: `Bearer ${token}`,
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify(post),
```
Comment (SaraHackle) [inline Tabloid/client/src/modules/postManager.js:66]: Consider adding error handling

File: Tabloid/client/src/components/PostEdit.js [hunk 3]
```diff
@@ -87,7 +93,10 @@ export const PostEdit = () => {
                         onInput={handleInputChange}
                     />
                 </FormGroup>
-                <Button disabled={!post.title || !post.content || !post.categoryId}>Submit</Button>
+                <Button disabled={!post.title || !post.content || !post.categoryId} onClick={(e) => updatePost()}>
+                    Save
+                </Button>
+                <Button onClick={(e) => history.push(`/posts`)}>Cancel</Button>
```
Comment (SaraHackle) [inline Tabloid/client/src/components/PostEdit.js]: Consider adding some space between buttons

**Review by SaraHackle — APPROVED** [review]
Looks great, my comments aren't absolutely necessary, so I approve!

File: Tabloid/client/src/modules/postManager.js (hunk 2 above)
Comment (cjgochanour) [reply Tabloid/client/src/modules/postManager.js:66]: considered

File: Tabloid/client/src/components/PostEdit.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    };
+
+    const updatePost = () => {
+        editPost(post).then(() => history.push(`/posts/${post.id}`));
+    };
+
+    useEffect(() => {
+        getPostById(id).then(setPost);
+        getAllCategories().then(setCategories);
+    }, []);
```
Comment (heymonicakay) [inline Tabloid/client/src/components/PostEdit.js]: This dependency array is empty right now, but it should probably watch for a change to the id variable.

File: Tabloid/client/src/components/PostEdit.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                    <Label for="dateInput">Date</Label>
+                    <Input
+                        id="publishDateTime"
+                        name="date"
+                        type="datetime-local"
+                        value={post.publishDateTime === null ? "" : post.publishDateTime || ""}
+                        onInput={handleInputChange}
+                    />
+                </FormGroup>
+                <Button disabled={!post.title || !post.content || !post.categoryId} onClick={(e) => updatePost()}>
```
Comment (heymonicakay) [inline Tabloid/client/src/components/PostEdit.js]: is the `e` in the parameter of the onclick callback necessary here?

File: Tabloid/client/src/components/PostEdit.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                    Save
+                </Button>
+                <Button className="ml-2" onClick={(e) => history.push(`/posts`)}>
```
Comment (heymonicakay) [inline Tabloid/client/src/components/PostEdit.js]: is the e in the parameter of the onclick callback necessary here?

**Review by heymonicakay — CHANGES_REQUESTED** [review]
Please see my comments

**Review by heymonicakay — APPROVED** [review]

**Pull request merged**